app/modules/uart.c

```diff
@@ -142,13 +142,27 @@ static int l_uart_write( lua_State* L )
   return 0;
 }
 
+static int l_uart_fifodepth( lua_State* L )
+{
+  int id  = luaL_optinteger( L, 1, 0 );
+  int dir = luaL_optinteger( L, 2, 0 );
+
+  int reg = READ_PERI_REG(UART_STATUS(id));
+  int rsh = reg >> (dir ? UART_TXFIFO_CNT_S : UART_RXFIFO_CNT_S);
+  int rm  = rsh &  (dir ? UART_TXFIFO_CNT   : UART_RXFIFO_CNT  );
+
```

Comment:
I wonder when this count goes to zero... Is it when the last byte is loaded into the UART shift register to be transmitted, or is it when the last byte has actually been transmitted? I don't think that there is much that can be done about this..... But it would nice to have a way of knowing that you could shut off the transmit pin ...

Reply:
The ESP8266 technical reference is not clear, saying only "Data has been sent will be automatically cleared, at the same time transmission FIFO will be more of a vacancy."

This makes it sound like a byte is removed from the FIFO only after it has been completely shifted out the pin, which would be nice, but it's more likely that it's just been moved into a shift register and is still being tx'd when the fifo depth is observed to be zero.  Tricky; hard to measure without writing some C, I think.  At 300 baud there might be enough time to get into the TX FIFO empty IRQ and toggle a GPIO to see what the waveform looks like on a 'scope...

Speaking of, should we expose the UART FIFO IRQs through `uart.on()`?